File: 1301-1400/1306-Jump Game III/1306-Jump Game III.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        if (arr[start] == 0) {
            return true;
        }
        unordered_set<int> visited;
        visited.insert(start);
        queue<int> Q;
        Q.push(start);
        while (!Q.empty()) {
            int curr = Q.front();
            Q.pop();
            for (int next : {curr + arr[curr], curr - arr[curr]}) {
                if (next >= 0 && next < arr.size()) {
                    if (arr[next] == 0) {
                        return true;
                    }
                    if (!visited.count(next)) {
                        visited.insert(next);
                        Q.push(next);
                    }
                }
            }
        }
        return false;
    }
};

File: 1301-1400/1306-Jump Game III/1306-Jump Game III.py
import collections
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        if arr[start] == 0:
            return True
        visited = set([start])
        Q = collections.deque([start])
        while Q:
            curr = Q.popleft()
            for next in (curr - arr[curr], curr + arr[curr]):
                if 0 <= next < len(arr):
                    if arr[next] == 0:
                        return True
                    if next not in visited:
                        visited.add(next)
                        Q.append(next)
        
        return False    
